chalk-engine/src/logic.rs

```diff
@@ -1288,6 +1301,60 @@ impl<'forest, C: Context + 'forest, CO: ContextOps<C> + 'forest> SolveState<'for
         }
     }
 
+    /// This strand has no subgoals left, but some have floundered.
+    /// It's still possible that we can get an ambiguous answer from it though
+    fn pursue_answer_from_floundered(&mut self, strand: Strand<C>) -> Option<AnswerIndex> {
+        let table = self.stack.top().table;
+        let Strand {
+            mut infer,
+            ex_clause:
+                ExClause {
+                    subst,
+                    constraints,
+                    ambiguous: _,
+                    subgoals,
+                    delayed_subgoals,
+                    answer_time: _,
+                    floundered_subgoals: _,
+                },
+            selected_subgoal: _,
+            last_pursued_time: _,
+        } = strand;
+        assert!(subgoals.is_empty());
+        let subst = infer.canonicalize_answer_subst(
+            self.context.interner(),
+            subst,
+            constraints,
+            delayed_subgoals,
+        );
+        debug!(
+            "answer from floundered goal: table={:?}, subst={:?}",
+            table, subst
+        );
+
+        let answer = Answer {
+            subst,
+            ambiguous: true,
+        };
+
+        // If our answer gives trivial information on the canonicalized goal then we have nothing interesting
+        // to return.
+        let is_trivial_answer = self
+            .context
+            .is_trivial_substitution(&self.forest.tables[table].table_goal, &answer.subst);
+
+        if is_trivial_answer {
+            None
```

Comment:
This is...strange. Maybe. An answer is an answer, even if it's ambiguous and trivial.

Reply:
Yeah that makes sense. I tried fixing it and I think it revealed a couple other issues (which I think are unrelated):
1. It's now possible that we will push an ambiguous answer and later push the same answer is unambiguous. This causes a panic at https://github.com/rust-lang/chalk/blob/80c17b8dfadde1440a2e4ddf5236bea650cef4f2/chalk-engine/src/table.rs#L138 where we assert that exactly that situation can't occur. There's a comment there saying that callers of `push_answer` assume a response of `true` doesn't mean an existing answer was updated, but removing the panic doesn't seem to cause any problems and the test that caused the panic now passes (`test::projection::iterator_flatten`).

1. In the test `test::misc::non_enumerable_traits_reorder` a subgoal flounders, then another one succeeds and that allows the first floundered subgoal to succeed. The test is now failing with the error:
```
expected:
substitution [?0 := Foo], lifetime constraints []
actual:
Ambiguous(substitution [?0 := Foo], lifetime constraints [])
```
So it seems like somehow we're not properly updating the existing answer from ambiguous to unambiguous (or maybe not even revisiting the floundered first subgoal?). I'll look into this one some more.

Reply:
This is actually great :) (In a totally "this is a bug that's been there that is getting uncovered" kind of a way).

So that first test `iterator_flatten`: I'm looking into it a bit more, but just removing the panic is *not* okay. This is a panic because we really need to invalidate *all* answers that depends on the ambiguous answer. This somewhat leads to the second point.

For the second, I think what we need to do is fix `make_solution` to be a little bit better at the guidance it creates. The problem is really that we create an initial ambiguous answer, then later create a non-ambiguous answer. We need to check that the non-ambiguous answer "covers" the ambiguous answer. And, if so, ignore that it's ambiguous. I'm not sure that we can just ignore all ambiguous answers all together, since they perhaps represent a different set of answers. It relates to the first problem somewhat in that *maybe* we just want to "restart" the aggregation.